src/compute/src/render/join/mz_join_core.rs

```diff
@@ -419,13 +420,23 @@ where
 
         let temp = &mut self.temp;
 
+        let flush = |data: &mut Vec<_>, session: &mut Session<_, _, _>| {
+            let old_len = data.len();
+            // Consolidating here is important when the join closure produces data that
+            // consolidates well, for example when projecting columns.
+            consolidate_updates(data);
+            let recovered = old_len - data.len();
+            session.give_iterator(data.drain(..));
```

Comment:
One question: We are now tying up consolidation with flushing, but perhaps we could have different policies for each? In particular, I think it is wise to consolidate right before flushing as above, but we could consolidate more often, e.g., every time the output buffer doubles to reduce the cost of consolidation.